src/java.base/share/classes/sun/security/provider/HSS.java

```diff
@@ -0,0 +1,837 @@
+            this.lmotsParams = lmotsParams;
+            n = lmotsParams.n;
+            int p = lmotsParams.p;
+            if (inLen != (4 + n * (p + 1))) {
+                throw new InvalidParameterException("OTS signature has incorrect length");
+            }
+            C = Arrays.copyOfRange(sigArray, 4, 4 + n);
+            int pStart = 4 + n;
+            y = new byte[p][n];
+            for (int i = 0; i < p; i++) {
+                y[i] = Arrays.copyOfRange(sigArray, pStart, pStart + n);
+                pStart += n;
+            }
+        }
+
+        void getC(byte[] arr, int pos) {
+            System.arraycopy(C, 0, arr, pos, n);
+        }
+
+        void getY(int i, byte[] arr, int pos) {
+            System.arraycopy(y[i], 0, arr, pos, n);
+        }
+    }
+
+    static class LMSParams {
+        final int m; // the number of bytes used from the hash output
+        final int hashAlg_m = 32; // output length of the LMS tree hash function
+        final int h; // height of the LMS tree
+        final int twoPowh;
+        final String hashAlgStr;
+
+        LMSParams(int m, int h, String hashAlgStr) {
+            this.m = m;
+            this.h = h;
+            this.hashAlgStr = hashAlgStr;
+            twoPowh = 1 << h;
+        }
+
+        static LMSParams of(int type) {
+            int m;
+            int h;
+            String hashAlgStr;
+            switch (type) {
+                case LMSUtils.LMS_SHA256_M32_H5:
+                    m = 32;
+                    h = 5;
+                    hashAlgStr = "SHA-256";
+                    break;
+                case LMSUtils.LMS_SHA256_M32_H10:
+                    m = 32;
+                    h = 10;
+                    hashAlgStr = "SHA-256";
+                    break;
+                case LMSUtils.LMS_SHA256_M32_H15:
+                    m = 32;
+                    h = 15;
+                    hashAlgStr = "SHA-256";
+                    break;
+                case LMSUtils.LMS_SHA256_M32_H20:
+                    m = 32;
+                    h = 20;
+                    hashAlgStr = "SHA-256";
+                    break;
+                case LMSUtils.LMS_SHA256_M32_H25:
+                    m = 32;
+                    h = 25;
+                    hashAlgStr = "SHA-256";
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unsupported or bad LMS type");
+            }
+
+            return new LMSParams(m, h, hashAlgStr);
+        }
+
+        boolean hasSameHash(LMSParams other) {
+            return other.hashAlgStr.equals(hashAlgStr) && (other.m == m);
+        }
+
+        boolean hasSameHash(LMOTSParams lmotsParams) {
+            return lmotsParams.hashAlgName.equals(hashAlgStr) &&
+                    (lmotsParams.n == m);
+        }
+    }
+
+    static class LMSignature {
+        final int sigLmType;
+        final int sigOtsType;
+        final private byte[] qArr;
```

Comment:
let's remove `private` to be consistent with other fields
```suggestion
        final byte[] qArr;
```

Reply:
I wanted to express that the array should be immutable.